### ISSUE TITLE
Automate as much of the "update imgui" process as is feasible

### ISSUE THREAD

**Issue by thomcc** [imgui-rs/imgui-rs#435]
The process right now is something like this (assuming [cimgui](https://github.com/cimgui/cimgui) is at `~/src/cimgui` and this repo is at `~/src/imgui-rs`):

1. Update [cimgui](https://github.com/cimgui/cimgui).
```sh
cd ~/src/cimgui
git pull
```

2. Update imgui submodule
```sh
cd ~/src/imgui-rs/imgui-sys/third-party/imgui
git remote update
git checkout <tag or whatever>
# Go back to imgui-rs root and probably commit
cd ../../..
```

3. Regenerate cimgui wrappers
```sh
cd imgui-sys/third-party
./update-cimgui-output.sh "$HOME/src/cimgui"
```

4. Regenerate cimgui wrappers
```sh
cd imgui-sys/third-party
./update-cimgui-output.sh "$HOME/src/cimgui"
```

5. (Optional) update build.rs and such for any changes, such as new cpp files.

6. Rerun bindgen
```sh
cd ~/src/imgui-rs
cargo xtask bindgen
```

7. Fix errors problems by new version and such.

8. Note the update in both the changelog, and bump the README.md version badge.

---

While I'd eventually like to replace the use of git submodule with [git subtree](https://www.atlassian.com/git/tutorials/git-subtree), it would be nice to automate this, as many of the steps could be automated. Doing this would make the transition to subtree easier, and allow a path forward for supporting more imgui branches.

In the mean time at least I wrote it down here so I can remember it next time. (A internal docs folder would be a better place for this, though)

**Comment by sanbox-irl** [comment]
You accidentally put `Regenerate cimgui wrappers` twice right?

**Comment by dbr** [comment]
Extremely useful to have that written down, thanks!

Maybe worth noting (if only for myself) that `~/src/cimgui/imgui` needs to be checked out to the same commit as `~/src/imgui-rs/imgui-sys/third-party/imgui` or things can go very wrong

**Comment by dbr** [comment]
As per my Friday update on #21 this is now done:

> # Friday
> 
>     * Made Rust segfault
> 
>     * imtodo functionality back to where it was on Monday
> 
> 
> ..and that's a good thing.
> 
> That is so say: I've managed to rebuilt the docking-branch support into `imgui-rs` from scratch, stumbling across and fixing lots of inevitable issues with binding to C++ libraries (things like if you share objects between them, they have to be the same size in memory - meaning structs have to be exactly correct size and order etc etc)
> 
> The process is actually quite simple - bit cumbersome and error prone, but now it's done once it should be fairly easy/quick to repeat as needed. Plus I learned quite a bit about how to work with bindgen etc (which will help greatly with future `imgui-rs` improvements we may need to make, particularly the freetype support)
> 
> Timing was quite nice - was almost about to give up for the day, then just around 6pm I worked out why things were crashing, added the one missing stuct field, and everything came together \o/
> 
> Incidentally the segfault was coming from this pointer-dereferencing inside `imgui-rs` (the `DrawList` struct was missing a field, so the pointer offset sizes were wrong size, thus, crash):
> 
> ```rust
> impl<'a> Iterator for DrawListIterator<'a> {
>     type Item = &'a DrawList;
> 
>     fn next(&mut self) -> Option<Self::Item> {
>         self.iter.next().map(|&ptr| unsafe { &*ptr })
>     }
> }
> ```
> 
> ..and that's exactly why Rust has `unsafe { ... }` blocks to lock away such crashy stuff

**Comment by sonoro1234** [comment]
> needs to be checked out to the same commit

I think that the best would be to have `~/src/imgui-rs/imgui-sys/third-party/cimgui` and use the imgui repo contained in cimgui

If the imgui version needs to be changed it can be done inside cimgui , then run cimgui-generator again and finally imgui-rs generator

**Comment by thomcc** [comment]
I don't agree. I find the submodule nesting causes problems for us, and it would make it harder for us to have different imgui branches enabled/disabled via cargo `features`. I'd rather move towards git subtree anyway.

**Comment by sonoro1234** [comment]
>  I find the submodule nesting causes problems for us


Of course you understand imgui-rs much better than me so I dont know which problems are caused.
In LuaJIT-ImGui I use cimgui the way told above. 


>  it would make it harder for us to have different imgui branches enabled/disabled via cargo `features`

I dont know anything about "cargo" so I cant tell. I always manage branches via github: each banch has a different cimgui submodule.


> I'd rather move towards git subtree anyway.

Not sure to understand.

**Comment by thomcc** [comment]
> I dont know anything about "cargo" so I cant tell. I always manage branches via github: each branch has a different cimgui submodule.

Yes, this would not allow us to allow different imgui branches as separate cargo features.

**Comment by sonoro1234** [comment]
Another example of having cimgui as a submodule is mellinoe/ImGui.NET where https://github.com/mellinoe/ImGui.NET-nativebuild
is used

**Comment by thomcc** [comment]
To elaborate a bit, to support what we'd want, we'd need one submodule for each branch, and you can see how this would get unwieldy.

**Comment by sonoro1234** [comment]
> we'd need one submodule for each branch

But this has the benefit that everything is synced without else to do.

**Comment by dbr** [comment]
This is documented in `docs/upgrading-imgui.md` (e.g [this one from 0.8 release](https://github.com/imgui-rs/imgui-rs/blob/v0.8.0/docs/upgrading-imgui.md))

https://github.com/imgui-rs/imgui-rs/pull/557 also adds an additional branch to generate things for - this is partially automated (`cargo xtask bindgen` runs bindgen for each branch), and the `update-cimgui-output.sh` script is a little simpler to use (automatically points it to correct source)... however it needs a bit more work, mainly to automate the checkout of imgui sources, and probably a wrapper to call all the update-cimgui scripts in one go

**Comment by dbr** [comment]
There's now a script to copy over imgui, along with the existing ones to run cimgui's generator and run bindgen (and the steps are documented)

I think this part is "automated enough" - probably takes less than a minute to do

The part which is a bit tedious and could maybe be more automated is when things change in big objects like `struct Io { ... }`. Finding the differences is really tedious and a bit time-consuming. It's basically just checking if all the structs are in the same order between the bindgen output and what is in `imgui/src/io.rs` (the tedium is because `imgui/src/io.rs` has doc-strings, slightly different names, and some slightly different types, so it requires a bit of manual reformatting to diff the two)
